### ISSUE TITLE
Fields not populating into template using Swift sample code

### ISSUE THREAD

**Issue by mjzandona** [docusign/native-ios-sdk#14]
I have followed the instructions to download the example app code, run the pod install, create template in developer sandbox, rename text box fields in Data Labels and use that name in the Swift code tabKey_* variables.  After following all of the instructions, the app is successfully built and running on the simulator.  However, all the text fields in the template show "Text" and not the data pre populated in the fields for Tom Woods.

![simulator](https://user-images.githubusercontent.com/4524664/55044447-95305580-4ff7-11e9-8cf1-b3e2c4944c8a.png)

I do see an error in the completionHandler in TemplatesManager.swift for the following call:

presentSendTemplateControllerWithTemplate

<img width="606" alt="error message" src="https://user-images.githubusercontent.com/4524664/55044614-40410f00-4ff8-11e9-87ce-e5c08dec5783.png">

I'm wondering if this error is causing the SDK to not properly execute and fill the fields in the call.

**Comment by iOSSDK-DocuSign** [comment]
Please ensure correct templateId is passed along with the tabData.

![template-id](https://user-images.githubusercontent.com/31968825/55095348-35788f80-5075-11e9-9422-b245801028f2.png)

-- Other relevant details -- 

Variables such as ‘tabKey_address2’ are used to link tabs in template and apply the default value before kicking off the signing ceremony. 

Line from swift sample:
`static let tabKey_address2: String = "Text 526bf465-583e-426f-9671-04dadbb16dbc";`

Tabs can be assigned a “Data Label” (tabLabel from API perspective).
For e.g. `Text 526bf465-583e-426f-9671-04dadbb16dbc` is the tabLabel (can be shared by multiple tabs).
 
Screenshot from DocuSign web with a template loaded and selected tab.
![tablabel-template](https://user-images.githubusercontent.com/31968825/55095574-9902bd00-5075-11e9-8194-95a4c1e88a26.png)

In order to pass a default value for a text-tab, you would need to identify the “Data Label” on the template and template-id, and use that string as the value.
 
`tabData["Text 526bf465-583e-426f-9671-04dadbb16dbc"] = “221 Main St Dallas TX“`

^ And this tabData defaults dictionary is used in sample app to invoke following: 
 
```
        // display template via sdk
        TemplatesManager.sharedInstance.displayTemplateForSignature(templateId: templateId, controller: self, tabData: tabData, recipientData: recipientData, customFields:customFields, onlineSign: false, attachmentUrl: attachmentUrl)
        { (errMsg) in
            
            // return to client screen
            let vcIndex = self.navigationController?.viewControllers.index(where: { (viewController) -> Bool in
                if let _ = viewController as? ClientViewController {
                    return true
                }
                return false
            })
            
            let clientVC = self.navigationController?.viewControllers[vcIndex!] as! ClientViewController
            self.navigationController?.popToViewController(clientVC, animated: true)
        }

```

Which in turn invokes sdk function to trigger the signing ceremony:

```
    func displayTemplateForSignature(templateId: String, controller: UIViewController, tabData: Dictionary<String, String>, recipientData: Array<DSMRecipientDefault>, customFields:DSMCustomFields?, onlineSign: Bool, attachmentUrl: URL?, completionHandler: @escaping ((UIViewController?, Error?) -> Void))
    {
        // load PDF data
        var pdfData: Data?
        if (attachmentUrl != nil)
        {
            do {
                pdfData = try Data(contentsOf: attachmentUrl!)
            }
            catch {
                NSLog("Error loading PDF data")
            }
        }
        
        let mDSMEnvelopeDefaults = DSMEnvelopeDefaults()
        mDSMEnvelopeDefaults.recipientDefaults = recipientData.count > 0 ? recipientData : nil
        mDSMEnvelopeDefaults.tabValueDefaults = tabData
        mDSMEnvelopeDefaults.customFields = customFields
        
        self.mDSMTemplatesManager?.presentSendTemplateControllerWithTemplate(withId: templateId, envelopeDefaults: mDSMEnvelopeDefaults, pdfToInsert: pdfData, insertAtPosition: DSMDocumentInsertAtPosition.end, signingMode: onlineSign ? DSMSigningMode.online : DSMSigningMode.offline, presenting: controller, animated: true, completion: completionHandler)
    }
```

**Comment by iOSSDK-DocuSign** [comment]
> I do see an error in the completionHandler in TemplatesManager.swift for the following call: presentSendTemplateControllerWithTemplate

Thanks for sharing this detail. Swift sample is out of date and we are planning to update it to 4.2.

**Comment by mjzandona** [comment]
Here are some screen shots showing you what I'm doing.  I'm following your examples as far as I can tell.

**Template ID passed to SDK:**
![templateID](https://user-images.githubusercontent.com/4524664/55111882-96fe2580-5098-11e9-8d77-b15f340f5841.png)

**Template ID in Xcode**
<img width="914" alt="templateIdXcode" src="https://user-images.githubusercontent.com/4524664/55111921-a7160500-5098-11e9-8a53-88ac61f7e4f4.png">

**tabData passed to SDK**
<img width="332" alt="tabData" src="https://user-images.githubusercontent.com/4524664/55111943-b301c700-5098-11e9-8b27-474f89ab00c8.png">

**template Data Label example**
![templateInfo](https://user-images.githubusercontent.com/4524664/55111958-bd23c580-5098-11e9-94a2-f72b95ac3067.png)

"U.S.A 94107" should be populating in my template, but all I see is "Text"

It keeps sounding like this example is very easy to implement.  But it simply isn't working for me.  Are there any other steps I could have missed that are preventing this from working properly?

**Comment by mjzandona** [comment]
Are you using a Powerform or a template?  From your screenshot I see a Recipient field which does not exist in my template form when I modify from the developer sandbox.
![Signer1](https://user-images.githubusercontent.com/4524664/55116015-345e5700-50a3-11e9-9942-c85f9b9562a1.png)

**Comment by iOSSDK-DocuSign** [comment]
This example is using templates. 

Here is a sample template `AutoInsuranceClaims.json` attached.

[AutoInsuranceClaims.json.zip](https://github.com/docusign/native-ios-sdk/files/3019351/AutoInsuranceClaims.json.zip)

Regarding not seeing the `recipient` field, it could be because your template only has 1 recipient? 

<img width="1053" alt="Screen Shot 2019-03-28 at 10 14 08 AM" src="https://user-images.githubusercontent.com/31968825/55179557-f6b20a80-5144-11e9-8976-031576e1731f.png">

**Comment by mjzandona** [comment]
I'm using DocuSign SDK v2.0, Xcode v10.1, Swift v3.

Can you run the swift example and provide a screenshot of the fields populated with the info from the app?  Let me know what version I should be running.

Also, if you have fixed the completion handler in the closure for the TemplateManager (it correctly has Viewcontroller and error as two passing arguments).    The DocusignViewController only has "errMsg" in the call to displayTemplateForSignature.

**Comment by mjzandona** [comment]
I see you made all the fixes/updates in v2.1 Swift example.  I will try starting from the beginning with your latest version.

**Comment by mjzandona** [comment]
I'm having problems installing the new pods

<img width="771" alt="podinstall" src="https://user-images.githubusercontent.com/4524664/55822638-e559ee00-5ab4-11e9-9fb6-095c1c1d984a.png">

Here is my directory structure

<img width="688" alt="directory" src="https://user-images.githubusercontent.com/4524664/55822651-ebe86580-5ab4-11e9-983f-d4d7868f4b2e.png">

**Comment by iOSSDK-DocuSign** [comment]
Please add `pod --version` and other details from the console output (top section) of `pod install --repo-update`.

**Comment by mjzandona** [comment]
Ok, got it now.  Project builds and executes but still same behavior as in my very first post above.  The data from the app does not get populated into the text fields.  

Is this user error of the app on my part?  I log in fine.  Click on Templates to cache my sample template.  Then I: 
- click Tom Woods (under upcoming appointments)
- click on 'View Agreement'
- click 'Next'
- click/select my sample template under TGK Advisor Agreements
- attach pdf
- click 'Next'
- click 'Sign'

At this point I'm presented my template with blue text boxes that are populated with "Text" instead of the info from the app.


Can you please build the sample app and template and show me a screen shot of the populated template?

**Comment by mjzandona** [comment]
I will wait for you to prove that the app works for you with a screenshot of a pre-populated template.  I'm going in circles.

When I debug the code, the tabData fields are getting properly assigned to the mDSMEnvelopeDefaults instance of the DSMEnvelopeDefaults class.  I should be seeing "Tom Wood" get populated into the fullname text field.

<img width="864" alt="code debug" src="https://user-images.githubusercontent.com/4524664/55831439-ec3e2c00-5ac7-11e9-8b99-f5519065d639.png">

**Comment by mjzandona** [comment]
I ended up solving this myself.  A little disappointed on the support.

I built the obj-c app and found that the template only gets populated properly in the online mode; offline mode does not work.
<img width="376" alt="AlluraFlow" src="https://user-images.githubusercontent.com/4524664/55933612-c6597a00-5be2-11e9-8269-dac2543eff4a.png">

Online Mode
<img width="378" alt="AlluraOnlineFlow" src="https://user-images.githubusercontent.com/4524664/55933624-d2453c00-5be2-11e9-9d76-2eb59f4712f0.png">

Offline mode
<img width="369" alt="AlluraOfflineFlow" src="https://user-images.githubusercontent.com/4524664/55933656-f274fb00-5be2-11e9-826e-6216b05b2cef.png">

The swift sample app defaults to offline mode.  I changed the logic so that it prompts the user to select the mode
<img width="681" alt="SwiftFlowCodeChange" src="https://user-images.githubusercontent.com/4524664/55933728-38ca5a00-5be3-11e9-8dda-7d56e27cd2c5.png">

and when I select online, I can get it to work.
<img width="359" alt="swift_working" src="https://user-images.githubusercontent.com/4524664/55933714-2c460180-5be3-11e9-81b7-13ac62d16f9c.png">

**Why does offline not work???**

**Comment by iOSSDK-DocuSign** [comment]
Hi @mjzandona,

In the screenshots you shared above for online you are using "Home Insurance Claim" template, that has the tab defaults data. 

In the offline mode, you are using "Auto Insurance Claim", that doesn't have tab defaults assigned during offline. 

In offline mode, If you use the same "Home Insurance Claim" template, you would see the tab defaults are also getting applied (exactly same data as online screenshot you shared). Here is the screenshot for your reference.

<img width="545" alt="Home_insurance_claim_adjuster_template-with-tab-defaults" src="https://user-images.githubusercontent.com/31968825/55982817-95af2a00-5c4e-11e9-918c-7e9bd4e8c64a.png">

Please find the the video and other files linked below to help understand the offline tab defaults use-case:

- [Templates & Screenshots](https://github.com/docusign/native-ios-sdk/blob/release/2.1.2/support-files/): Home Insurance (with tab defaults) & Auto Insurance (without tab defaults)
- Video: [Offline Signing walkthrough](https://github.com/docusign/native-ios-sdk/blob/release/2.1.2/support-files/offline-signing-with-and-without-tab-defaults.mov) with and without tab defaults. 
- **Note:** These files have been uploaded to latest release (2.1.2) branch.

**Comment by mjzandona** [comment]
The first set of 3 screen shots is from the obj-c sample app. And the only option is to use the auto insurance claim.  All the screen shots are using auto insurance. I did not use home insurance template.   Now I’m confused. The movie would be helpful but it looks like the file is too big. Can you host it on Dropbox or YouTube?

On Apr 11, 2019, at 11:50 AM, iOSSDK-DocuSign <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @mjzandona<https://github.com/mjzandona>,

In the screenshots you shared above for online you are using "Home Insurance Claim" template, that has the tab defaults data.

In the offline mode, you are using "Auto Insurance Claim", that doesn't have tab defaults assigned during offline.

In offline mode, If you use the same "Home Insurance Claim" template, you would see the tab defaults are also getting applied (exactly same data as online screenshot you shared). Here is the screenshot for your reference.

[Home_insurance_claim_adjuster_template-with-tab-defaults][55982817-95af2a00-5c4e-11e9-918c-7e9bd4e8c64a.png]<https://user-images.githubusercontent.com/31968825/55982817-95af2a00-5c4e-11e9-918c-7e9bd4e8c64a.png>

Please find the the video and other files linked below to help understand the offline tab defaults use-case:

  *   Templates & Screenshots<https://github.com/docusign/native-ios-sdk/blob/release/2.1.2/support-files/>: Home Insurance (with tab defaults) & Auto Insurance (without tab defaults)
  *   Video: Offline Signing walkthrough<https://github.com/docusign/native-ios-sdk/blob/release/2.1.2/support-files/offline-signing-with-and-without-tab-defaults.mov> with and without tab defaults.
  *   Note: These files have been uploaded to latest release (2.1.2) branch.

**Comment by mjzandona** [comment]
Your screen shot of "home insurance claim" from the obj-c app is actually the "investment agreement" from the swift sample app.  It's like we're talking about different sample apps.

In the insurance app, clicking on home insurance claim does nothing (it's a dummy option).  The only option that does anything is the auto insurance option which triggers the sample template provided by you that gets added to my sandbox.

**Comment by mjzandona** [comment]
Just watched your video.  What is different between the Auto Insurance and Home Insurance templates you are selecting in the swift app?

**Comment by iOSSDK-DocuSign** [comment]
Regarding Objective-C sample: 

You should be able to access and use both templates, "Home Insurance" & "Auto Insurance", on Objective-C sample as well with offline. 

<img width="545" alt="ObjC-Allura-accessing-both-templates" src="https://user-images.githubusercontent.com/31968825/55992193-d9129400-5c60-11e9-8197-6761d702f683.png">

Please have a look at the video for [Offline signing walkthrough with Objective-C based](https://github.com/docusign/native-ios-sdk/blob/release/2.1.2/support-files/objc-offline-signing-with-and-without-tab-defaults.mov) sample app. 

In order to access the template in offline mode, user would need to download it first. Once the template has been cached on device for offline use, user may start the offline signing session with additional recipients and/or tab default values. Both samples also show adding an extra PDF documents to append with the template documents during signing.

**Comment by mjzandona** [comment]
Thanks.  More specifically,  what is different for each template in your developer sandbox?  How is the signer area configured?

<img width="849" alt="recipient" src="https://user-images.githubusercontent.com/4524664/55993582-5095f280-5c64-11e9-8e72-4698c2a3c465.png">

The sample repository has one pdf file that gets uploaded to the developer sandbox to create the template (you are showing two templates in the app: Home and Auto.  What is the difference because the same pdf file is being used?

I add the text fields, copy the Data Label "variable" names into the swift code.  I assume that is the same for both of your templates.

**Comment by mjzandona** [comment]
Is there a way to chat and screenshare with you?  I think that would simplify all this back and forth.  I'm not sure if we're on the same page.

**Comment by iOSSDK-DocuSign** [comment]
> Thanks. More specifically, what is different for each template in your developer sandbox? 

Each of the templates contain different configurations (signers, documents, etc). For details on templates, please take a look at the Official DocuSign support page:
- [Working with Templates](https://support.docusign.com/guides/ndse-user-guide-working-with-templates)
- There are a bunch of helpful videos such as [What are DocuSign Templates?](https://www.youtube.com/watch?v=6yU_mIyFZPE) linked there. 


> How is the signer area configured?

To configure the signer area within [Templates: Create Templates](https://docs.docusign.com/DocuSignHelp/Content/create-templates.htm). 

> Is there a way to chat and screenshare with you? I think that would simplify all this back and forth. I'm not sure if we're on the same page.

Yes, It would be great to have someone get in touch with you. Working on getting someone from customer support or developer evangelism team to get help on these matters. Will get in touch with you over email.

> The sample repository has one pdf file that gets uploaded to the developer sandbox to create the template (you are showing two templates in the app: Home and Auto. What is the difference because the same pdf file is being used?

"Home Insurance" and "Auto Insurance" are using different PDFs. They look quite similar, but they are different, one has the title "Investment Agreement" and other has "Car Accident Claim" as title. 

![auto-insurance](https://user-images.githubusercontent.com/31968825/55996143-635ff580-5c6b-11e9-8188-06fbdc6ff5e7.png)
![home-insurance](https://user-images.githubusercontent.com/31968825/55996144-635ff580-5c6b-11e9-920f-954a9360bc6a.png)


> I add the text fields, copy the Data Label "variable" names into the swift code. I assume that is the same for both of your templates.

You would also need to make sure that `Data Labels` for both templates match. And correct `templateId` is used. You would need to edit the `Data Labels` manually to match (screenshot added in first reply to this issue/thread).

**Comment by mjzandona** [comment]
Ok thanks. 

In your swift movie, your "**_Home Insurance_**" template (which is really the **_Investment Agreement_**) is the one that works properly in offline mode.  If you could show me how your signer area is configured, I can probably figure this out on my own.  My problem is that I get various errors in the app when caching depending on how I setup the template signer/host information in the sandbox.  This is the only way I can get "in person signer" to work is the following:

<img width="849" alt="recipient" src="https://user-images.githubusercontent.com/4524664/55997036-32cd8b00-5c6e-11e9-9cce-47c465cf59d2.png">

If I change any of the fields I get an error in the app that says:

<img width="298" alt="error" src="https://user-images.githubusercontent.com/4524664/55997111-64465680-5c6e-11e9-938c-adad1f9fa672.png">

None of this is an issue when I configure my template for online mode.  Everything works.  The only downside is that you don't support pdf insertion for online mode.  Can that feature be requested?  What is the reason not to allow adding pdf docs to an envelope in online signing mode?

**Comment by mjzandona** [comment]
Is it possible I don't have proper permission to use "offline" mode?  Why would the same exact template work for online mode (in person signing) but not work for offline mode.  Offline mode seems to be a special feature that may require a special license?

I will try to post a video that shows me using the same exact template for online and offline signing.  Offline is not populating the information.

**Comment by iOSSDK-DocuSign** [comment]
Yes. You would need to have offline signing & templates enabled in your account. Getting you in touch with the customer support team.

<img width="849" alt="55997036-32cd8b00-5c6e-11e9-9cce-47c465cf59d2" src="https://user-images.githubusercontent.com/31968825/56063865-530c5100-5d25-11e9-85e6-f2adcd311561.png">

**Comment by mjzandona** [comment]
Thank you. It feels like I'm finally getting somewhere.

For my application, I don't require offline signing.  But I do need ability to have pdf insert to envelopes.  

From the documentation, it looks like pdf insert is ignored for online signing.  Why is that the case?

<img width="982" alt="offlinesigning" src="https://user-images.githubusercontent.com/4524664/56068839-3841d880-5d35-11e9-8e55-94400ce66b5f.png">

**Comment by iOSSDK-DocuSign** [comment]
Hi Michael, 

> From the documentation, it looks like pdf insert is ignored for online signing. Why is that the case?
This is something that's not supported out of the box for online signing. 

For v2.x, alternative is to use offline signing flow to create envelopes with templates and add additional  pdf with `pdfToInsert` parameter and proceed with offline signing. Once the signing is complete in offline mode, app/user may sync the envelope back to DocuSign servers when the network is reachable.

Also, sent you a meeting invitation.

**Comment by iOSSDK-DocuSign** [comment]
Got this resolved on side channel. This is the updated code in swift sample app that  

```
    /**
     Returns Array of the recipient data that will be passed into the template
     */
    func getTemplateRecipientData(templateId:String) -> Array<DSMRecipientDefault>
    {
        // comment out following to fix the recipient default not getting applied to template on an external developer account 
        /*if templateId != ProfileManager.Static.templateIdMomemtumDemo && templateId != ProfileManager.Static.templateIdMomemtumDemoWireTransfer {
            NSLog("Warning: Recipient Data not found for given templateId. Returning empty data.")
            return []
        }*/
        
        let recipientDatum = DSMRecipientDefault.init()
        // Use recipient roleName (other option to use recipient-id) to find unique recipient in the template
        recipientDatum.recipientRoleName = "claimant-roleName"
        recipientDatum.recipientSelectorType = DSMEnvelopeDefaultsUniqueRecipientSelectorType.recipientRoleName
        recipientDatum.recipientType = DSMRecipientType.inPersonSigner
        // In-person-signer name
        recipientDatum.inPersonSignerName = "Tom Wood"
        // Host name (must match the name on the account) and email
        recipientDatum.recipientName = "docusignsdk user"
        recipientDatum.recipientEmail = "docusignsdk.user@dsxtr.com"
        return [recipientDatum]
    }
```

Will push this change with upcoming release.

**Comment by ashokds** [comment]
@mjzandona , closing this as resolved.

**Comment by pushp1010** [comment]
@ashokds Hi Ashok, 

I tried building my Swift sample Where I am able to login but when I tried to set Tab data but it does not works for me.

below is the screenshot of the Text Id I am using. I am not sure what I am I doing wrong. based on the Documentation and my knowledge it should work.

<img width="979" alt="Screenshot 2021-05-04 at 6 30 23 PM" src="https://user-images.githubusercontent.com/52121407/117007113-dfbc4a00-ad06-11eb-8cc7-3de57acc4b1e.png">

Below is the code snippet: 

                
                print(template.templateId)
                let recipientDatum = DSMRecipientDefault()
                // Use recipient roleName (other option to use recipient-id) to find unique recipient in the template
                recipientDatum.recipientRoleName = "Lead"
                recipientDatum.recipientSelectorType = .recipientRoleName
                recipientDatum.recipientType = .inPersonSigner
                // In-person-signer name
                recipientDatum.inPersonSignerName = "Pushpendra Singh"
                // Host name (must match the name on the account) and email
                recipientDatum.recipientName = "pushpendra singh"
                recipientDatum.recipientEmail = "psinghiosdeveloper@gmail.com"
                
                let textCustomField = DSMTextCustomField()
                textCustomField.name = "Text 06a6bafd-991f-438c-8df2-2f13dc114609"
                textCustomField.value = "Tom Wood"
                textCustomField.show = true

                
                let tabData1: [String: String] = ["Text 06a6bafd-991f-438c-8df2-2f13dc114609": "jas"]
                
                let customFields = DSMCustomFields()
                customFields.textCustomFields = [textCustomField]
                
                TemplatesManager.sharedInstance.displayTemplateForSignature(templateId: template.templateId, controller: self.viewController!, tabData: tabData1, recipientData: [recipientDatum], customFields: customFields, onlineSign: true, attachmentUrl: nil) { (controller, error) in
                    
                    print(controller, error)
                }


I am trying onlineSign It should work. Please put on some light here.

**Comment by ashokds** [comment]
@pushp1010 , could you please open a new issue and link with this ( https://github.com/docusign/native-ios-sdk/issues/14 )
